File: ios/Classes/FlutterPollfishPlugin.m
#import "FlutterPollfishPlugin.h"
#import <Pollfish/Pollfish-Swift.h>

FlutterMethodChannel *_channel_pollfish;

@implementation FlutterPollfishPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    _channel_pollfish = [FlutterMethodChannel
                                     methodChannelWithName:@"flutter_pollfish"
                                     binaryMessenger:[registrar messenger]];
    
 
    FlutterPollfishPlugin* instance = [[FlutterPollfishPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:_channel_pollfish];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {

    if ([@"init" isEqualToString:call.method]) {
        [self callInitialize:call result:result];
        return;
    } else if ([@"show" isEqualToString:call.method]) {
        [Pollfish show];
        return;
    } else if ([@"hide" isEqualToString:call.method]) {
        [Pollfish hide];
        return;
    } else if ([@"isPollfishPresent" isEqualToString: call.method]) {
        result([NSNumber numberWithBool:[Pollfish isPollfishPresent]]);
        return;
    } else if ([@"isPollfishPanelOpen" isEqualToString: call.method]) {
        result([NSNumber numberWithBool:[Pollfish isPollfishPanelOpen]]);
        return;
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)callInitialize:(FlutterMethodCall *) call result:(FlutterResult) result {
    NSString *apiKey = (NSString *) call.arguments[@"iOSApiKey"];
    
    if (apiKey == [NSNull null] || [apiKey length] == 0) {
        result([FlutterError errorWithCode:@"no_api_key"
                                   message:@"a non-empty Pollfish API Key was not provided"
                                   details:nil]);
        return;
    }
    
    PollfishParams *params = [[PollfishParams alloc] init:apiKey];
    
    if (call.arguments[@"indicatorPosition"] != [NSNull null]) {
        int indicatorPosition = (int) [call.arguments[@"indicatorPosition"] integerValue];
        [params indicatorPosition: (IndicatorPosition) indicatorPosition];
    }

    if (call.arguments[@"indicatorPadding"] != [NSNull null]) {
        int indicatorPadding = (int) [call.arguments[@"indicatorPadding"] integerValue];
        [params indicatorPadding: indicatorPadding];
    }

    if (call.arguments[@"releaseMode"] != [NSNull null]) {
        BOOL releaseMode = (BOOL) [call.arguments [@"releaseMode"] boolValue];
        [params releaseMode: releaseMode];
    }

    if (call.arguments[@"rewardMode"] != [NSNull null]) {
        BOOL rewardMode = (BOOL)[call.arguments[@"rewardMode"] boolValue];
        [params rewardMode: rewardMode];
    }

    if (call.arguments[@"requestUUID"] != [NSNull null]) {
        NSString *requestUUID = call.arguments[@"requestUUID"];
        [params requestUUID: requestUUID];
    }

    if (call.arguments[@"offerwallMode"] != [NSNull null]) {
        BOOL offerwallMode = (BOOL)[call.arguments[@"offerwallMode"] boolValue];
        [params offerwallMode: offerwallMode];
    }
    
    if (call.arguments[@"userProperties"] != [NSNull null]) {
        NSDictionary *userProperties = (NSDictionary *) call.arguments[@"userProperties"];
        UserProperties *userPropertiesBuilder = [[UserProperties alloc] init];
            
        [userProperties enumerateKeysAndObjectsUsingBlock:^(id key, id object, BOOL *stop) {
            [userPropertiesBuilder customAttribute:object forKey:key];
        }];
        
        [params userProperties:userPropertiesBuilder];
    }
    
    if (call.arguments[@"clickId"] != [NSNull null]) {
        NSString *clickId = call.arguments[@"clickId"];
        [params clickId:clickId];
    }
    
    if (call.arguments[@"signature"] != [NSNull null]) {
        NSString *signature = call.arguments[@"signature"];
        [params signature:signature];
    }
    
    if (call.arguments[@"rewardInfo"] != [NSNull null]) {
        NSDictionary *rewardInfoDict = (NSDictionary *) call.arguments[@"rewardInfo"];
        
        NSString *rewardName = [rewardInfoDict objectForKey:@"rewardName"];
        double rewardConversion = [[rewardInfoDict objectForKey:@"rewardConversion"] doubleValue];
        
        if (rewardName != nil) {
            RewardInfo *rewardInfo = [[RewardInfo alloc] initWithRewardName:rewardName rewardConversion:rewardConversion];
            [params rewardInfo:rewardInfo];
        }
    }

    [params platform: PlatformFlutter];
    
    [Pollfish initWith:params delegate: self];
}

- (NSDictionary *)getDictionaryFrom:(SurveyInfo *) surveyInfo
{
    if (surveyInfo == nil) {
        return nil;
    }
    
    NSDictionary * payload = @{
        @"surveyCPA" : (surveyInfo.cpa ?: [NSNull null]),
        @"surveyIR" : (surveyInfo.ir ?: [NSNull null]),
        @"surveyLOI" : (surveyInfo.loi ?: [NSNull null]),
        @"rewardValue" : (surveyInfo.rewardValue ?: [NSNull null]),
        @"remainingCompletes" : (surveyInfo.remainingCompletes ?: [NSNull null]),
        @"surveyClass" : (surveyInfo.surveyClass ?: [NSNull null]),
        @"rewardName" : (surveyInfo.rewardName ?: [NSNull null])};
    
    return payload;
}

- (void) pollfishSurveyCompletedWithSurveyInfo:(SurveyInfo *)surveyInfo
{
    [_channel_pollfish invokeMethod:@"pollfishSurveyCompleted"
                          arguments:[self getDictionaryFrom:surveyInfo]];
}

- (void)pollfishOpened
{
     [_channel_pollfish invokeMethod:@"pollfishOpened" arguments:nil];
}

- (void)pollfishClosed
{
    [_channel_pollfish invokeMethod:@"pollfishClosed" arguments:nil];
}

- (void) pollfishSurveyReceivedWithSurveyInfo:(SurveyInfo *)surveyInfo
{
    [_channel_pollfish invokeMethod:@"pollfishSurveyReceived"
                          arguments:[self getDictionaryFrom:surveyInfo]];
}

- (void)pollfishUserNotEligible
{
    [_channel_pollfish invokeMethod:@"pollfishUserNotEligible" arguments:nil];
}

- (void)pollfishUserRejectedSurvey
{
    [_channel_pollfish invokeMethod:@"pollfishUserRejectedSurvey" arguments:nil];
}

- (void) pollfishSurveyNotAvailable
{
    [_channel_pollfish invokeMethod:@"pollfishSurveyNotAvailable" arguments:nil];
}

@end

File: android/src/main/java/com/pollfish/flutterpollfish/FlutterPollfishPlugin.java
package com.pollfish.flutterpollfish;

import android.app.Activity;
import android.util.Log;

import androidx.annotation.NonNull;

import com.pollfish.Pollfish;
import com.pollfish.builder.Params;
import com.pollfish.builder.Platform;
import com.pollfish.builder.Position;
import com.pollfish.builder.RewardInfo;
import com.pollfish.builder.UserProperties;
import com.pollfish.callback.PollfishClosedListener;
import com.pollfish.callback.PollfishOpenedListener;
import com.pollfish.callback.PollfishSurveyCompletedListener;
import com.pollfish.callback.PollfishSurveyNotAvailableListener;
import com.pollfish.callback.PollfishSurveyReceivedListener;
import com.pollfish.callback.PollfishUserNotEligibleListener;
import com.pollfish.callback.PollfishUserRejectedSurveyListener;
import com.pollfish.callback.SurveyInfo;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;


/**
 * FlutterPollfishPlugin
 */
public class FlutterPollfishPlugin implements FlutterPlugin, ActivityAware, MethodCallHandler {

    public static final String TAG = "FlutterPollfishPlugin";

    private MethodChannel channel;
    private FlutterPluginBinding binding = null;
    private Activity activity = null;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
        channel = new MethodChannel(binding.getBinaryMessenger(), "flutter_pollfish");
        channel.setMethodCallHandler(this);
        this.binding = binding;
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
        this.binding = null;
    }

    @Override
    public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
        activity = binding.getActivity();
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {
    }

    @Override
    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
    }

    @Override
    public void onDetachedFromActivity() {
        activity = null;
    }

    private Map<String, Object> getMapFromSurveyInfo(SurveyInfo surveyInfo) {
        if (surveyInfo == null) {
            return null;
        }

        Map<String, Object> payload = new HashMap<>();

        if (surveyInfo.getSurveyCPA() != null) {
            payload.put("surveyCPA", surveyInfo.getSurveyCPA());
        }

        if (surveyInfo.getSurveyLOI() != null) {
            payload.put("surveyLOI", surveyInfo.getSurveyLOI());
        }

        if (surveyInfo.getSurveyIR() != null) {
            payload.put("surveyIR", surveyInfo.getSurveyIR());
        }

        if (surveyInfo.getSurveyClass() != null) {
            payload.put("surveyClass", surveyInfo.getSurveyClass());
        }

        if (surveyInfo.getRewardName() != null) {
            payload.put("rewardName", surveyInfo.getRewardName());
        }

        if (surveyInfo.getRewardValue() != null) {
            payload.put("rewardValue", surveyInfo.getRewardValue());
        }

        if (surveyInfo.getRemainingCompletes() != null) {
            payload.put("remainingCompletes", surveyInfo.getRemainingCompletes());
        }

        return payload;
    }

    private void initPollfish(final Activity activity,
                              final String apiKey,
                              final int indicatorPosition,
                              final int indicatorPadding,
                              final boolean releaseMode,
                              final boolean rewardMode,
                              final boolean offerwallMode,
                              final String requestUUID,
                              Map<String, Object> userProperties,
                              String clickId,
                              String userId,
                              String signature,
                              Map<String, Object> rewardInfo) {

        final Position position = Position.values()[indicatorPosition];

        Params.Builder paramsBuilder = new Params.Builder(apiKey)
                .indicatorPosition(position)
                .indicatorPadding(indicatorPadding)
                .pollfishSurveyCompletedListener(new PollfishSurveyCompletedListener() {
                    @Override
                    public void onPollfishSurveyCompleted(@NotNull SurveyInfo surveyInfo) {
                        channel.invokeMethod("pollfishSurveyCompleted", getMapFromSurveyInfo(surveyInfo));
                    }
                })
                .pollfishSurveyReceivedListener(new PollfishSurveyReceivedListener() {
                    @Override
                    public void onPollfishSurveyReceived(@Nullable SurveyInfo surveyInfo) {
                        channel.invokeMethod("pollfishSurveyReceived", getMapFromSurveyInfo(surveyInfo));
                    }
                })
                .pollfishClosedListener(new PollfishClosedListener() {
                    @Override
                    public void onPollfishClosed() {
                        channel.invokeMethod("pollfishClosed", null);
                    }
                })
                .pollfishOpenedListener(new PollfishOpenedListener() {
                    @Override
                    public void onPollfishOpened() {
                        channel.invokeMethod("pollfishOpened", null);
                    }
                })
                .pollfishUserNotEligibleListener(new PollfishUserNotEligibleListener() {
                    @Override
                    public void onUserNotEligible() {
                        channel.invokeMethod("pollfishUserNotEligible", null);
                    }
                })
                .pollfishUserRejectedSurveyListener(new PollfishUserRejectedSurveyListener() {
                    @Override
                    public void onUserRejectedSurvey() {
                        channel.invokeMethod("pollfishUserRejectedSurvey", null);
                    }
                })
                .pollfishSurveyNotAvailableListener(new PollfishSurveyNotAvailableListener() {
                    @Override
                    public void onPollfishSurveyNotAvailable() {
                        channel.invokeMethod("pollfishSurveyNotAvailable", null);
                    }
                })
                .releaseMode(releaseMode)
                .rewardMode(rewardMode)
                .platform(Platform.FLUTTER)
                .offerwallMode(offerwallMode);

        if (userProperties != null && !userProperties.isEmpty()) {
            UserProperties.Builder userPropertiesBuilder = new UserProperties.Builder();

            try {
                for (String s : userProperties.keySet()) {
                    if (userProperties.get(s) != null && userProperties.get(s) instanceof String) {
                        userPropertiesBuilder.customAttribute(s, (String) userProperties.get(s));
                    }
                }
            } catch (Exception e) {
                Log.e(TAG, "Error extracting userProperties: " + e);
            }

            paramsBuilder.userProperties(userPropertiesBuilder.build());
        }

        if (rewardInfo != null && !rewardInfo.isEmpty()) {
            RewardInfo pollfishRewardInfo = null;

            try {
                String rewardName = (String) rewardInfo.get("rewardName");
                Double rewardConversion = (Double) rewardInfo.get("rewardConversion");

                if (rewardName != null && rewardConversion != null) {
                    pollfishRewardInfo = new RewardInfo(rewardName, rewardConversion);
                }
            } catch (Exception e) {
                Log.e(TAG, "Error extracting rewardInfo: " + e);
            }

            if (pollfishRewardInfo != null) {
                paramsBuilder.rewardInfo(pollfishRewardInfo);
            }
        }

        if (requestUUID != null) {
            if (requestUUID.trim().length() > 0) {
                paramsBuilder.requestUUID(requestUUID);
            }
        }

        if (clickId != null) {
            if (clickId.trim().length() > 0) {
                paramsBuilder.clickId(clickId);
            }
        }

        if (userId != null) {
            if (userId.trim().length() > 0) {
                paramsBuilder.userId(userId);
            }
        }

        if (signature != null) {
            if (signature.trim().length() > 0) {
                paramsBuilder.signature(signature);
            }
        }

        Pollfish.initWith(activity, paramsBuilder.build());
    }

    private void exctractPollfishParams(Activity activity, MethodCall call, Result result) {
        String apiKey = null;

        if (call.argument("androidApiKey") != null) {
            apiKey = call.argument("androidApiKey");
        }

        if (apiKey == null) {
            result.error("no_api_key", "a null apiKey was provided", null);
            return;
        }

        int pollfishPosition = 0;
        int indicatorPadding = 50;
        boolean releaseMode = false;
        boolean rewardMode = false;
        boolean offerwallMode = false;
        String requestUUID = null;
        Map<String, Object> userProperties = null;
        String clickId = null;
        String userId = null;
        String signature = null;
        Map<String, Object> rewardInfo = null;

        if (call.argument("indicatorPosition") != null) {
            pollfishPosition = call.argument("indicatorPosition");
        }

        if (call.argument("indicatorPadding") != null) {
            indicatorPadding = call.argument("indicatorPadding");
        }

        if (call.argument("releaseMode") != null) {
            releaseMode = call.argument("releaseMode");
        }

        if (call.argument("rewardMode") != null) {
            rewardMode = call.argument("rewardMode");
        }

        if (call.argument("offerwallMode") != null) {
            offerwallMode = call.argument("offerwallMode");
        }

        if (call.argument("requestUUID") != null) {
            requestUUID = call.argument("requestUUID");
        }

        if (call.argument("userProperties") != null) {
            userProperties = call.argument("userProperties");
        }

        if (call.argument("clickId") != null) {
            clickId = call.argument("clickId");
        }

        if (call.argument("userId") != null) {
            userId = call.argument("userId");
        }

        if (call.argument("signature") != null) {
            signature = call.argument("signature");
        }

        if (call.argument("rewardInfo") != null) {
            rewardInfo = call.argument("rewardInfo");
        }

        if (binding != null)
            initPollfish(activity, apiKey, pollfishPosition, indicatorPadding, releaseMode, rewardMode, offerwallMode, requestUUID, userProperties, clickId, userId, signature, rewardInfo);

        result.success(null);
    }

    @Override
    public void onMethodCall(MethodCall call, @NotNull Result result) {
        switch (call.method) {
            case "init":
                if (activity != null) {
                    exctractPollfishParams(activity, call, result);
                }
                break;
            case "show":
                Pollfish.show();
                break;
            case "hide":
                Pollfish.hide();
                break;
            case "isPollfishPresent":
                result.success(Pollfish.isPollfishPresent());
                break;
            case "isPollfishPanelOpen":
                result.success(Pollfish.isPollfishPanelOpen());
                break;
            default:
                result.notImplemented();
        }
    }
}
